modules/miner/update.go

```diff
@@ -28,11 +27,138 @@ func (m *Miner) addNewTxns(diff *modules.TransactionPoolDiff) {
 				totalFees = totalFees.Add(fee)
 			}
 		}
-		m.splitSets[m.setCounter] = &splitSet{
+		// We will check to see if this splitSet belongs in the block.
+		s := &splitSet{
 			size:         size,
 			averageFee:   totalFees.Div64(size),
 			transactions: newSet.Transactions,
 		}
+
+		elem := &mapElement{
+			set:   s,
+			id:    splitSetID(m.setCounter),
+			index: 0,
+		}
+
+		m.splitSets[splitSetID(m.setCounter)] = s
+		newElements = append(newElements, elem)
+	}
+	return newElements
+}
+
+// addMapElementTxns places the splitSet from a mapElement into the correct
+// mapHeap.
+func (m *Miner) addMapElementTxns(elem *mapElement) {
+	candidateSet := elem.set
+
+	// Check if heap for highest fee transactions has space.
+	if m.blockMapHeap.size+candidateSet.size < types.BlockSizeLimit-5e3 {
+		m.blockMapHeap.Push(elem)
+		return
+	}
+
+	// While the heap cannot fit this set s, and while the (weighted) average
+	// fee for the lowest sets from the block is less than the fee for the set
+	// s, continue removing from the heap. The block heap doesn't have enough
+	// space for this transaction. Check if removing sets from the blockMapHeap
+	// will be worth it. bottomSets will hold  the lowest fee sets from the
+	// blockMapHeap
+	bottomSets := make([]*mapElement, 0)
+	var sizeOfBottomSets uint64
+	var averageFeeOfBottomSets types.Currency
+	for {
+		// Check if the candidateSet can fit in the block.
+		if m.blockMapHeap.size-sizeOfBottomSets+candidateSet.size < types.BlockSizeLimit-5e3 {
+			// Place candidate into block,
+			m.blockMapHeap.Push(elem)
+
+			// Places transactions removed from block heap into
+			// the overflow heap.
+			for _, v := range bottomSets {
+				m.overflowMapHeap.Push(v)
+			}
+			break
+		}
+
+		// If the blockMapHeap is empty, push all elements removed from it back
+		// in, and place the candidate set into the overflow. This should never
+		// happen since transaction sets are much smaller than the max block
+		// size.
+		_, exists := m.blockMapHeap.Peek()
+		if !exists {
+			m.overflowMapHeap.Push(elem)
+			// Put back in transactions removed.
+			for _, v := range bottomSets {
+				m.blockMapHeap.Push(v)
+			}
+			// Go to the next candidate set.
```

Comment:
comment is confusing because the function has no context of multiple candidate sets, all it knows is that it is finishing up the current candidate set.